File: android/src/com/android/tools/idea/model/IdeBuildType.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.model;

import com.android.annotations.Nullable;
import com.android.builder.model.BuildType;
import com.android.builder.model.SigningConfig;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;

/**
 * Creates a deep copy of {@link BuildType}.
 *
 * @see IdeAndroidProject
 */
public class IdeBuildType extends IdeBaseConfig implements BuildType, Serializable {
  @Nullable private final SigningConfig mySigningConfig;
  private final boolean myDebuggable;
  private final boolean myTestCoverageEnabled;
  private final boolean myPseudoLocalesEnabled;
  private final boolean myJniDebuggable;
  private final boolean myRenderscriptDebuggable;
  private final int myRenderscriptOptimLevel;
  private final boolean myMinifyEnabled;
  private final boolean myZipAlignEnabled;
  private final boolean myEmbedMicroApp;

  public IdeBuildType(@NotNull BuildType type) {
    super(type);

    SigningConfig tySigningConfig = type.getSigningConfig();
    mySigningConfig = tySigningConfig == null ? null : new IdeSigningConfig(tySigningConfig);

    myDebuggable = type.isDebuggable();
    myTestCoverageEnabled = type.isTestCoverageEnabled();
    myPseudoLocalesEnabled = type.isPseudoLocalesEnabled();
    myJniDebuggable = type.isJniDebuggable();
    myRenderscriptDebuggable = type.isRenderscriptDebuggable();
    myRenderscriptOptimLevel = type.getRenderscriptOptimLevel();
    myMinifyEnabled = type.isMinifyEnabled();
    myZipAlignEnabled = type.isZipAlignEnabled();
    myEmbedMicroApp = type.isEmbedMicroApp();
  }

  @Override
  @Nullable
  public SigningConfig getSigningConfig() {
    return mySigningConfig;
  }

  @Override
  public boolean isDebuggable() {
    return myDebuggable;
  }

  @Override
  public boolean isTestCoverageEnabled() {
    return myTestCoverageEnabled;
  }

  @Override
  public boolean isPseudoLocalesEnabled() {
    return myPseudoLocalesEnabled;
  }

  @Override
  public boolean isJniDebuggable() {
    return myJniDebuggable;
  }

  @Override
  public boolean isRenderscriptDebuggable() {
    return myRenderscriptDebuggable;
  }

  @Override
  public int getRenderscriptOptimLevel() {
    return myRenderscriptOptimLevel;
  }

  @Override
  public boolean isMinifyEnabled() {
    return myMinifyEnabled;
  }

  @Override
  public boolean isZipAlignEnabled() {
    return myZipAlignEnabled;
  }

  @Override
  public boolean isEmbedMicroApp() {
    return myEmbedMicroApp;
  }
}

File: project-system-gradle-models/src/com/android/tools/idea/gradle/model/IdeBuildType.kt
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.gradle.model

interface IdeBuildType : IdeBaseConfig {
  /** Whether the build type is configured to generate a debuggable apk. */
  val isDebuggable: Boolean

  /** Whether the build type is configured to generate an apk with debuggable native code. */
  val isJniDebuggable: Boolean

  /** Whether the build type is configured to generate an apk with debuggable renderscript code. */
  val isRenderscriptDebuggable: Boolean

  /** The optimization level of the renderscript compilation. */
  val renderscriptOptimLevel: Int

  /**
   * Specifies whether to enable code shrinking for this build type.
   *
   * By default, when you enable code shrinking by setting this property to `true`,
   * the Android plugin uses ProGuard.
   *
   * To learn more, read
   * [Shrink Your Code and Resources](https://developer.android.com/studio/build/shrink-code.html).
   */
  val isMinifyEnabled: Boolean

  /** Whether zipalign is enabled for this build type. */
  val isZipAlignEnabled: Boolean

  /** Whether this build type is specified as a default by the user*/
  val isDefault: Boolean?
}
